torch/testing/_internal/common_utils.py

```diff
@@ -832,219 +939,241 @@ def safeCoalesce(self, t):
+        torch.complex64  : (1.3e-6, 1e-5),
+        torch.complex128 : (1e-7, 1e-7),
     }
 
-    # todo: implement numpy-like issubdtype
-    def is_integral(self, dtype):
-        # Skip complex/quantized types
-        dtypes = [x for x in get_all_dtypes() if x not in get_all_complex_dtypes()]
-        return dtype in dtypes and not dtype.is_floating_point
-
-    # accepts tensors, dtypes, or np.ndarrays
-    def get_default_tolerance(self, a, b=None):
-        if b is None:
-            dtype = torch.float
-            if isinstance(a, torch.Tensor):
-                dtype = a.dtype
-            elif isinstance(a, torch.dtype):
-                dtype = a
-            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
-                # Some tests call assertEqual with numpy Unicode.
-                if numpy.issubdtype(a.dtype, numpy.dtype('U')):
-                    dtype = torch.float
-                else:
-                    dtype = torch.from_numpy(a).dtype
-            if self.is_integral(dtype):
-                return (0, 0)
-            dtype = str(dtype).split('.')[-1]
-            return self.dtype_precisions.get(dtype, (self.precision, self.precision))
-
-        a_tol = self.get_default_tolerance(a)
-        b_tol = self.get_default_tolerance(b)
-        return (max(a_tol[0], b_tol[0]), max(a_tol[1], b_tol[1]))
-
-    def assertEqual(self, x, y, message='', *, atol=None, rtol=None, allow_inf=False, exact_dtype=None):
+    # Checks if two dense tensors are equal(-ish), returning (True, None)
+    # when they are and (False, debug_msg) when they are not, where the
+    # debug_msg is a string containing information about why the tensors
+    # failed to compare as equal(-ish).
+    # If exact_dtype is true both tensors must have the same dtype.
+    # If exact_device is true both tensors must be on the same device.
+    # See NOTE Test Framework Tensor "Equality" for more information on
+    # how tensors are compared.
+    # NOTE: tensors on different devices are moved to the CPU to be compared when
+    # exact_device is False.
+    def _compareTensors(self, a, b, *, rtol=None, atol=None, equal_nan=True,
+                        exact_dtype=False, exact_device=False):
+        if not isinstance(a, torch.Tensor):
+            return (False, "argument a, {0}, to _compareTensors is not a tensor!".format(a))
+        if not isinstance(b, torch.Tensor):
+            return (False, "argument b, {0}, to _compareTensors is not a tensor!".format(b))
+
+        # Validates tensors are on the same device
+        if exact_device and a.device != b.device:
+            return (False, ("Attempted to compare equality of tensors on "
+                            "different devices! Got devices {0} and "
+                            "{1}.".format(a.device, b.device)))
+
+        # Compares tensors of different devices on the CPU
+        if a.device != b.device:
+            a = a.cpu()
+            b = b.cpu()
```

Comment:
not sure if this matters, but in master it's `b=b.to(a)`, meaning if b is on cuda it would stay on cuda.

Reply:
I know. I thought this was simpler to understand and more consistent. Cross-device comparisons are rare and very few tests should rely on assertEqual performing them. Even from a cursory review I noticed several tests incorrectly relying on assertEqual to move tensors across CUDA devices. In the future we should default exact_device to True to better address this.